mathjax3-ts/output/chtml/Wrappers/mtable.ts

```diff
@@ -0,0 +1,387 @@
+        }
+    }
+
+    /*
+     * Add borders to the tops of cells to make the row lines
+     */
+    protected handleRowLines() {
+        if (this.node.attributes.get('rowlines') === 'none') return;
+        const lines = this.getRowAttributes('rowlines');
+        if (!lines) return;
+        let i = 0;
+        for (const row of this.childNodes.slice(1)) {
+            const line = lines[i++];
+            if (line === 'none') continue;
+            for (const cell of Array.from(row.chtml.childNodes) as HTMLElement[]) {
+                cell.style.borderTop = '.07em ' + line;
+            }
+        }
+    }
+
+    /*
+     * Add a frame to the mjx-itable, if needed
+     */
+    protected handleFrame(frame: boolean) {
+        if (frame) {
+            (this.chtml.firstChild as HTMLElement).style.border = '.07em ' + this.node.attributes.get('frame');
+        }
+    }
+
+    /******************************************************************/
+
+    /*
+     * @param{string} name           The name of the attribute to get as an array
+     * @param{CHTMLWrapper} wrapper  The wrapper whose attribute is to be used
+     * @return{string[]}             The array of values in the given attribute, split at spaces,
+     *                                 padded to the number of table columns by repeating the last entry
+     */
+    protected getColumnAttributes(name: string, wrapper: CHTMLWrapper = null) {
+        const columns = this.getAttributeArray(name, wrapper);
+        if (columns.length === 0) return;
+        while (columns.length < this.numCols) {
+            columns.push(columns[columns.length - 1]);
+        }
+        return columns;
+    }
+
+    /*
+     * @param{string} name           The name of the attribute to get as an array
+     * @param{CHTMLWrapper} wrapper  The wrapper whose attribute is to be used
+     * @return{string[]}             The array of values in the given attribute, split at spaces,
+     *                                 padded to the number of table rows by repeating the last entry
+     */
+    protected getRowAttributes(name: string, wrapper: CHTMLWrapper = null) {
+        const rows = this.getAttributeArray(name, wrapper);
+        if (rows.length === 0) return;
+        while (rows.length < this.numRows) {
+            rows.push(rows[rows.length - 1]);
+        }
+        return rows;
+    }
+
+    /*
+     * @param{string} name           The name of the attribute to get as an array
+     * @param{CHTMLWrapper} wrapper  The wrapper whose attribute is to be used
+     * @return{string[]}             The array of values in the given attribute, split at spaces
+     *                                 (after leading and trailing spaces are removed, and multiple
+     *                                  spaces have been collapsed to one).
+     */
+    protected getAttributeArray(name: string, wrapper: CHTMLWrapper = null) {
+        const value = ((wrapper || this).node.attributes.get(name) as string);
```

Comment:
I am surprised this type checks...

Reply:
Well, `this` is a CHTMLWrapper and so is `wrapper`, so `(wrapper || this)` is a CHTMLWrapper, and those have `node` properties, so that should be OK.  The attributes are `string | number | boolean` as I recall, so casting to `string` should be OK.